### ISSUE TITLE
`%to(addressbook,other.custom*)%` not working from smart fragments menu [Cardbook]

### ISSUE THREAD

**Issue by orcofthedarkness** [RealRaven2000/SmartTemplates#246]
Hi,

TB version: 115.0
SmartTemplates version: 4.0pre256
Cardbook version: 90.4

Here is the content of my template (it doesn't matter "which" template (write, reply, forward...)
`<p>Hi %to(firstname)%<br /><br />Dein User: %to(addressbook,other.custom1)%</p>`

When you have **NO** license applied, both the _%to(firstname)%_ and the _%to(addressbook,other.custom1)%_ are working just fine.
Once you apply the (valid) license (all PRO features are enabled), the %to(firstname)% still works, but _%to(addressbook,other.custom1)%_ **doesn't work anymore**. If you remove the license key, everything work as expected.

Tested also with other.custom[2-4].

thx in advance,
Ron

**Comment by RealRaven2000** [comment]
Could be a compatibility bug in Cardbook
That you get a result without license suggests that the data was retrieved from the standard AB instead. Do you see any error messages in the error console (open with Ctrl_Shift+J) when you lload the template?

**Comment by orcofthedarkness** [comment]
yes, i get the following 2 messages, when i click on the clickable variable
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/e983733c-2f83-4daf-80ad-740aa3d5d161)

**Comment by RealRaven2000** [comment]
Ok. could you export one contact for me for testing please? Please send to thunderbirddaily@gmail.com, please reference [SmartTemplates] issue 246 in the subject line,  if you want to send it off Github.

Also, I while testing I discovered a bad issue #247 that breaks the templates file in all versions of SmartTemplates 4.0, so don't try to add templates to the menu until that is fixed (is going to go into the next beta with high priority)

**Comment by RealRaven2000** [comment]
Thanks for sending test data - I couldn't reproduce the particular p[roblem (TypRror: nodeList is undefined), I got partial data back but didn't get the data for "other1, other2":

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/b3def926-bb0e-4f7a-b437-19dc87f8023e)

It looks like in my previous test versions I would have expected a arrauy member looking like this:

"VALUE:some string"

but we see this data in card.others (information partially redacted):

```
card.others
Array [ "X-CUSTOM1:WolfgangRaffetseder", "X-CUSTOM2:r*******w" ]
0: "X-CUSTOM1:WolfgangR*******"​
1: "X-CUSTOM2:r*******w"​
length: 2

```

**Comment by RealRaven2000** [comment]
Below is a test template for all Cardbook / Addressbook fields

[Address-Book-tests.zip](https://github.com/RealRaven2000/SmartTemplates/files/12064915/Address-Book-tests.zip)

**Comment by RealRaven2000** [comment]
The error of nodeList = null seems to point to the way you are using the field. Interactively? from a plain text message? If the account template contains the address book fields, can you please export your smartTemplates account templates : too?

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/82db4610-9f3b-4865-81d4-49ab9301f5aa)

(this includes all the data on the left side of the dialog for the selected account)

**Comment by orcofthedarkness** [comment]
Here is the exported account json
[Ronald.zip](https://github.com/RealRaven2000/SmartTemplates/files/12064948/Ronald.zip)

**Comment by orcofthedarkness** [comment]
> Below is a test template for all Cardbook / Addressbook fields
> 
> [Address-Book-tests.zip](https://github.com/RealRaven2000/SmartTemplates/files/12064915/Address-Book-tests.zip)

now it gets weird.
I installed the attached pre273 version from https://github.com/RealRaven2000/SmartTemplates/issues/247
When i import the templates from the "Address-Book-test.zip" as  HTML template:  Write - the correct data is pulled from the AB
When i import the templates as Smart Snippet, it doesnt work any longer, no data will be pulled from the AB

**Comment by RealRaven2000** [comment]
> > Below is a test template for all Cardbook / Addressbook fields
> > [Address-Book-tests.zip](https://github.com/RealRaven2000/SmartTemplates/files/12064915/Address-Book-tests.zip)
> 
> now it gets weird. I installed the attached pre273 version from #247 When i import the templates from the "Address-Book-test.zip" as HTML template: Write - the correct data is pulled from the AB When i import the templates as Smart Snippet, it doesnt work any longer, no data will be pulled from the AB

great that helps. I am inserting HTML elements  into the document and they are later resolved in form of a list. Maybe that fails with a document that's pulled from the smartTemplate menu. For more info in error console, please install the version below

[smartTemplate-fx-4.0pre274.zip](https://github.com/RealRaven2000/SmartTemplates/files/12064990/smartTemplate-fx-4.0pre274.zip)

this will avoid the nodeList problem but *not* do anything immediately (e.g. try to insert text into the Editor immediately). I need to test first, but it should advance us a little...

**Comment by orcofthedarkness** [comment]
installed pre274, used my template as a template for new mail

1) write new email -> template is used in msg body
2) fill in receipient in TO field
3) click on the clickable variables (one is working, on is not (see [above](https://github.com/RealRaven2000/SmartTemplates/issues/246#issue-1806649779))

Heres the output:
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/6b688309-7c6b-4c1f-ac8f-8956a09c59c9)

**Comment by RealRaven2000** [comment]
Personally, I wouldn't ever use `%to%` (or `%from%`) in a snippet. Because it is not clear who that is. The to from the original email that you reply to or the to you are now writing to? 

instead, use either `%recipient()%` (for whoever you are sending the email to) or `%identity()%` representing the sender of this email (your from identity). To and from are very confusing to use for the text of the template, but are ok to use within quote header.

try this instead:

```html
<p>Hi %recipient(firstname)%<br /><br />
Dein User: %recipient(addressbook,name,other.custom1,other.custom2)%
</p>
```

**Comment by orcofthedarkness** [comment]
> Personally, I wouldn't ever use `%to%` (or `%from%`) in a snippet. Because it is not clear who that is. The to from the original email that you reply to or the to you are now writing to?

yes, that makes sense. thanks. changed it to your suggested version but still...no luck :(
template used as template for a new email (and click ond the clickable variables)
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/bc2524ee-951f-4752-a203-85fb4eee3865)

template used as smart snippet: doesnt work at all

**Comment by RealRaven2000** [comment]
By hte way, testing with THunderbird 102, here is what I get returned from a custom1 field:

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/c039d7e4-8259-4c14-b477-6f29d9d3a858)

the first line is the name of the custom property, the second line is the contained data. That's why I was looking for "VALUE:(some data)". So I need to check why your data is not formatted that way ... hope it is not vCard version dependant, I based it on the format that Phillipe V (the owner of the  Cardbook Add-on) returned at the time I implemented the communication between ST and CB

**Comment by RealRaven2000** [comment]
I am still down the rabbit hole of testing. it seems there are some "mixed up perspective" problems still when it tries to get the header variables. The function `classifyReservedWord(..)` inthis case uses lookup list TokenMap["recipient"] and that returns "to"... which is correct, but the "hdr" is filled with the header information of the original mail (the one I reply to). I wonder if this is an artifact of using the addressbook lookup during a fragment, I still have to wrap my head around that.

**Comment by RealRaven2000** [comment]
Ok I made some unfortunate finding what happens when we use the fragment in a reply. For `%recipient%` it doesn't use the current value from the "to" address but the "from" of the original email header. So if we change the addressee by typing a completely different recipient in the "to:" address this will be ignored and still retrieve data from the original sender. I guess I need to look very hard whether I can either swap out the "hdr" structure to reflect data from the current document once the fragments are used, or not replace the "recipient" keywords until much later. Depending on what is easier; this is complicated stuff right there...

**Comment by RealRaven2000** [comment]
I dealt with the %recipient% problem, at least it works in fragments now. I will very likely raise a separate bug to explain it. Here is a version for testing:

[smartTemplate-fx-4.0pre277.zip](https://github.com/RealRaven2000/SmartTemplates/files/12065286/smartTemplate-fx-4.0pre277.zip)

the upshot is that you can now edit the addresse and change the "To:" and the fragments will pick up the correct address items via `%recipient()%`.

**Comment by orcofthedarkness** [comment]
> I dealt with the `%recipient%` problem, at least it works in fragments now. I will very likely raise a separate bug to explain it. Here is a version for testing:
> 
> [smartTemplate-fx-4.0pre277.zip](https://github.com/RealRaven2000/SmartTemplates/files/12065286/smartTemplate-fx-4.0pre277.zip)

tested, looks good to me (re-populate the clickable variables after changing the receipient in TO: field)

one more thing (not related to the pre277 version): if i use the html template as snipplet, i cannot populate cthe clickable variables. if i use the very same template in a "new e-mail":
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/a0137929-bd7f-4fe8-821b-2f67a28ea80f)

here you can see, that its the same template (i dont have a template in "write email":
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/850027c4-5514-4f78-afc1-ef6848ab7253)

**Comment by RealRaven2000** [comment]
> if i use the html template as snipplet, i cannot populate the clickable variables.

Sorry I don't understand that sentence, can you give an example? 

Do you mean the "deferred" variables that turn up when you write a new email using fields like %recipient(name)% ?

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/a79e62cf-408f-4aea-a150-4388b093e902)

**Comment by orcofthedarkness** [comment]
> > if i use the html template as snipplet, i cannot populate the clickable variables.
> 
> Sorry I don't understand that sentence, can you give an example?
> 
> Do you mean the "deferred" variables that turn up when you write a new email using fields like %recipient(name)% ?
> 
> ![image](https://user-images.githubusercontent.com/6314674/254639413-a79e62cf-408f-4aea-a150-4388b093e902.png)

sorry...my bad. my text was little confusing.
My template starts like this:

`<p>Hi %recipient(firstname)%,<br /><br />Dein User: %recipient(addressbook,name,other.custom1)%</br>Dein Passwort: %clipboard%</p>`

if i insert this template via
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/11a4ed28-3f69-40f8-a1be-bf235055428e)

and click on the first "To - button" (next to "Hi") in the msg body, nothing happens at all.
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/d2e88d5a-89af-4490-8606-10042349f0f1)

if i insert the very same template via 
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/e0515b1a-c80d-41d8-8a70-a94f259c1250)

a click on the first "To - button" (next to "Hi") works indeed
![image](https://github.com/RealRaven2000/SmartTemplates/assets/127893294/806139bc-f325-46f4-902f-a7135c60f78b)

**Comment by RealRaven2000** [comment]
I am not sure if I can make the script commands work from a "Fragments" entry - did this use to work in Thunderbird 102? There may also be some fallout from the changes (fixes) to %recipient%. I mainly use the "insert snippet" menu for extracting information from the original mail or adding some FAQ / Layout elements, nothing too heavy that interacts with the address headers .... 

I made a new version over at:

https://github.com/RealRaven2000/SmartTemplates/issues/249#issuecomment-1642668168

Are there any errors in javascript error console? (use the right-click menu to copy all messages, no screenshots please)

**Comment by RealRaven2000** [comment]
Since we closed #249 can we close this one as well? I am not sure if there any open issues... if it still is a problem, simply reopen this one and send the error log...

**Comment by RealRaven2000** [comment]
> I am just revisiting this bug to make sure it is fixed - I don't know where I can edit the "custom" values in the CardBook editor, any ideas?

I think I managed it in cardbook settings under Custom fields, I set up this:

![image](https://github.com/RealRaven2000/SmartTemplates/assets/6314674/c7248634-408b-46c9-adc9-32663f1b83e1)

Then I made a new version which makes sure to use the correct email when using the variables `%recipient%` and `%identity%` even from a smart fragment menu and used the following fragment for testing:

```html
<div style="border: 1px solid gray; padding: 0.5em; margin-block: 0.5em; border-radius: 5px;">
  <h3>test cardbook <i>other.custom</i>:</h3> 
  Recipient: %recipient(addressbook,name)%<br>
  [[C1: %recipient(addressbook,other.custom1)%<br>]]
  [[Shoe Size: %recipient(addressbook,other.custom2)%<br>]]
  [[YouTuber: %recipient(addressbook,other.custom3)%<br>]]
  [[Support: %recipient(addressbook,other.custom4)%<br>]]
</div>
```

The results looked correct to me - test version below:

[smartTemplate-fx-4.4pre17.zip](https://github.com/RealRaven2000/SmartTemplates/files/14334319/smartTemplate-fx-4.4pre17.zip)

---

To test the version above, download the zip file and drag the file into Thunderbird Add-ons Manager **without extracting contents**.

**Comment by RealRaven2000** [comment]
Implemented in 4.4 - published 28/03/2024
